### PR TITLE
Stop removed node and start added node 

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
#306 
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: curp/src/server/raw_curp/mod.rs [hunk 1]
```diff
@@ -551,22 +553,43 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
...
+                .iter()
+                .any(|(_, ctx)| match ctx.origin_entry.entry_data {
+                    EntryData::ConfChange(ref cc) => cc.changes().iter().any(|c| {
+                        matches!(c.change_type(), ConfChangeType::Remove)
+                            && c.node_id == candidate_id
+                    }),
+                    EntryData::Empty(_) | EntryData::Command(_) | EntryData::Shutdown(_) => false,
+                });
+        // extra check to shutdown removed node
+        if !contains_candidate && !remove_candidate_is_not_committed {
```
Comment (bsbds) [inline curp/src/server/raw_curp/mod.rs:569]: Should this check be added to handle_vote? It it possible that a removed leader complete a pre-vote and start a vote?

File: curp/src/server/raw_curp/mod.rs [hunk 2]
```diff
@@ -618,6 +641,8 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
 
         let prev_last_log_index = log_w.last_log_index();
         self.recover_from_spec_pools(&mut st_w, &mut log_w, spec_pools);
+        let propose_id = ProposeId(rand::random(), 0);
```
Comment (bsbds) [inline curp/src/server/raw_curp/mod.rs:644]: Should this id generation use the same way of the client id refactor(client id need to be acquired from server). If so please add a todo.

File: curp/src/server/raw_curp/mod.rs [hunk 3]
```diff
@@ -893,7 +900,10 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
             lst_r.term
         };
 
-        let next_index = self.lst.get_next_index(follower_id);
+        let Some(next_index) = self.lst.get_next_index(follower_id) else {
+            warn!("follower {} is not found, it maybe has been removed", follower_id);
```
Comment (bsbds) [inline curp/src/server/raw_curp/mod.rs:904]: If the follower has been removed, please make sure that the leader stop all the sync task.

File: curp/tests/common/curp_group.rs [hunk 4]
```diff
@@ -164,19 +163,92 @@ impl CurpGroup {
         debug!("successfully start group");
         Self {
             nodes,
-            all,
             storage_path,
         }
     }
 
+    pub async fn run_node(
```
Comment (bsbds) [inline curp/tests/common/curp_group.rs:170]: What's the main difference between `run_node` and `new` in CurpGroup?

File: curp/src/server/raw_curp/mod.rs (hunk 3 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:904]: all the sync task? Do you mean the sync task of the removed node? the task will be stop when this method return `None`

File: curp/tests/common/curp_group.rs (hunk 4 above)
Comment (themanforfree) [reply curp/tests/common/curp_group.rs:170]: `new` will create an initialized cluster, `run_node` will run a new node and insert to current cluster

File: curp/src/server/raw_curp/mod.rs (hunk 2 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:644]: Done

File: curp/src/server/raw_curp/mod.rs (hunk 1 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:569]: If the pre-vote is complete, there is no healthy leader in the current cluster, so it is impossible to process `MemberRemoveRequest`.

File: simulation/tests/it/curp/server_recovery.rs [hunk 5]
```diff
@@ -47,15 +47,18 @@ async fn leader_crash_and_recovery() {
     group.restart(leader, false).await;
     let old_leader = group.nodes.get_mut(&leader).unwrap();
 
+    // new leader will push a empty log to commit previous logs, the empty log does
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply simulation/tests/it/curp/server_recovery.rs]: Fixed

File: curp/tests/server.rs [hunk 6]
```diff
@@ -588,18 +575,18 @@ async fn shutdown_rpc_should_shutdown_the_cluster_when_client_has_wrong_cluster(
...
+    // init_logger();
     let tmp_path = tempfile::TempDir::new().unwrap().into_path();
-    let group = CurpGroup::new_rocks(3, tmp_path.clone()).await;
+    let mut group = CurpGroup::new_rocks(3, tmp_path.clone()).await;
     let client = group.new_client().await;
 
     let listener = TcpListener::bind("0.0.0.0:0").await.unwrap();
     let addrs = vec![listener.local_addr().unwrap().to_string()];
     let node_id = ClusterInfo::calculate_member_id(addrs.clone(), "", Some(123));
+    println!("node_id: {}", node_id);
```
Comment (bsbds) [inline curp/tests/server.rs:589]: Please remove this debug print.

File: curp/tests/server.rs [hunk 7]
```diff
@@ -578,14 +578,15 @@ async fn shutdown_rpc_should_shutdown_the_cluster_when_client_has_wrong_cluster(
 #[tokio::test(flavor = "multi_thread")]
 #[abort_on_panic]
 async fn propose_conf_change_rpc_should_work_when_client_has_wrong_cluster() {
-    init_logger();
+    // init_logger();
```
Comment (Phoenix500526) [inline curp/tests/server.rs:581]: Why do you comment this `init_logger`?

Comment (themanforfree) [reply curp/tests/server.rs]: Fixed

File: curp/tests/server.rs (hunk 6 above)
Comment (themanforfree) [reply curp/tests/server.rs]: Removed

**Review by Phoenix500526 — APPROVED** [review]
LGTM. Please squash commits.

**Review by bsbds — APPROVED** [review]

**Pull request merged**